src/systems/filecoin_markets/storage_market/storage_market_actor.go

```diff
@@ -312,6 +317,44 @@ func (a *StorageMarketActorCode_I) GetLastDealExpirationFromDealIDs(rt Runtime,
 	return lastDealExpiration
 }
 
+func (a *StorageMarketActorCode_I) GetUnsealedCIDForDealIDs(rt Runtime, sectorSize util.UVarint, dealIDs []deal.DealID) vmr.InvocOutput {
```

Comment:
this souldn't be here in the `StorageMarketActor` -- this is chain code in the VM. this should be in the `SectorBuilder` i think. See the `SectorBuilder.NextSector()` function call (doesnt have code yet), i think this belongs there

Reply:
I'm not sure I follow. At the point that we are verifying a Seal proof, we need access to the deals which were actually put on chain — in order to determine whether or not the submitted proof actually corresponds to the claimed deals.

Reply:
moving potentially large objects across vm boundaries is costly (serialization cost and allocations for that must be paid, you can't pass in-memory objects between actors). Pulling all of these out of the store only to shove them in an array that gets serialized, thrown off to another actor, and then unmarshaled back into that array, only to be reduced to a single hash immediately, is unnecessary.

We should work hard to make everything in the deal flow as low cost as possible, its already going to be tight, don't need to make things worse on ourselves.

We should strive to not have any methods that require potentially large complex structures to move between actors.

Additionally, I can see how this method is very relevant to the storage market, this can be framed as a helper that lets you easily check if your deal is part of a bigger hunk of data.

Reply:
> (serialization cost and allocations for that must be paid, you can't pass in-memory objects between actors).

i think this is an implementation problem? The requirement is a safe memory model, that ensures no mutable memory is shared across actor calls to ensure no changes are made. but that doesn't require serialization on the spot. (probably don't want to do that, will be really expensive, especially for cases where calls fail or run out of gas, which will need to be reverted as a whole. (dont want to hit disk every time)

For example, an implementation with functional, immutable data structures and good caching would do this very cheaply and **_safely_**, by passing references to immutable in-memory data structures with lazy serialization, or to CIDs that load quickly because the object is kept in cache until the VM invocation is finished

> We should work hard to make everything in the deal flow as low cost as possible, its already going to be tight, don't need to make things worse on ourselves.

Very much agree. And we need to not make things worse on ourselves in terms of protocol correctness and verification. Implementing and verifying all of the code is drastically easier when modules are well scoped and concerns don't leak through